docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1995 @@
+{
```

Comment:
> In this case, I would take the `and min_dist >= 1e-6` out of the while condition and add a break inside the while loop before we update `min_dist` :

So, we are allowing `min_dist` to become lower than 1e-6, which is set in `config.STUMPY_MIN_DISCORD`? Correct? 
This can be considered as edge case so we do not need to worry about it. 

**edge case (Just for the record)** 
if  `min_dist` is larger than but close to `1e-6`... then... we skip `if-break block` and update(reduce) `min_dist`. And thus the new value of `min_dist` might become lower than `1e-6`. One solution might be:

```
if min_dist <= 1e-6:  #note equal/smaller sign
     break

min_dist = min_dist * (1 - decay[len(discords_idx)])
min_dist = max(min_dist, 1e-6)
```
So, if `min_dist` becomes less than 1e-6...we reset it to 1e-6... we do one last check for discords with `min_dist=1e-6`, and
then we break out of while-loop.

---

> Hmmm, there seems to be a misnomer in `min_dist` . The docstring says:
> 
> > The lowest z-normalized Euclidean distance that a discord should have with ALL of its neighbors.
> 
> But it appears that we don't actually do anything to enforce that the minimum distance is greater than or equal to what the user defined. Technically, before we return `out` back to the user, we should be comparing it to the user-defined `min_dist` to ensure that it meets their criteri

I agree about the misleading definition in docstring. I think we need to change the docstring... and help reader understand the concept behind `min_dist`. We can think of `min_dist` as **initial guess** (that is why we lower the threshold whenever we cannot find a discord). What matters here is `k` (number of discords) requested by user.  

**I THINK what you are trying to say here is similar to `max_distance` in `stumpy.motifs`.** Am I right? In `motifs`, you make sure the  distances of motifs and their neighbors do not exceed `max_distance`.  **I am trying to understand your suggestion when you said:** 
```
if min_dist is not None: 
# Remove all discords where the distance is smaller than min_dist
```
**And, I think something is not right here.** So, let us assume user set `min_dist=10`. We then do: `min_D = min_dist` (so initial value of `min_D` is 10) and we use `min_D` in the while-loop. Correct? We search for discords that have `distance>=min_D`. Then, after finding all discords with `dist>=min_D`, we try to lower `min_D` to find more discords. (I am considering `else` in while-loop) 

NOTE THAT, after updating `min_D`, the new discords have distance lower than 10 (initial value of min_D). Why? Because, if their distance was  at least min_D, they would be discovered already in previous iterations! In other words, you will end up throwing all these discords (the ones discovered  after updating `min_D`)

We might change the name min_dist to `min_dist_final`... i.e. the lowest distance a discord allowed to have with its NearestNeighbor.
```
if min_dist_final is None:
    min_dist_final = config.STUMPY_MIN_DISCORD
```
In other words, if the user do not care about distance of discord to its NearestNeighbor, and they just want to get top-k discords anyway, we are allowing `min_dist` to be reduced up to ` min_dist_final=config.STUMPY_MIN_DISCORD`. 
So, we can then set `min_dist` ourselves (e.g. `min_dist = 2 * \sqrt(m)`). We want to start with large `min_dist` to avoid getting LOTs of candidates and LOTs of discords. We reduce `min_dist` in while-loop and we stop when `min_dist` becomes lower than `min_dist_final`.

Reply:
<p>Regarding the edge case, yes, I agree with what you proposed. We should set <code>min_dist = 1e-6</code> if it ever drops below <code>config.STUMPY_MIN_DISCORD_DISTANCE</code> </p><p> </p><blockquote><strong>I THINK what you are trying to say here is similar to&nbsp;</strong><code><strong>max_distance</strong></code><strong>&nbsp;in&nbsp;</strong><code><strong>stumpy.motifs</strong></code><strong>.</strong>&nbsp;Am I right? In&nbsp;<code>motifs</code>, you make sure the distances of motifs and their neighbors do not exceed&nbsp;<code>max_distance</code></blockquote><p> </p><p>Yes. I'm starting to wonder if we should try to be consistent with the paper and simply call this <code>r</code> since <code>min_dist</code> makes it too ambiguous.  And then borrow from the description in section B of the paper.</p><p> </p><p>Please ignore what I said about remove discords that are greater than <code>min_dist</code>.  Due to the name and the docstring, I was confused. I think renaming it to <code>r</code> will be better. It won't be as descriptive but it will at least be consistent with the paper.</p><p> </p><p>  </p><p>Regarding <code>n = 0</code> , I just meant something like:</p><p> </p><pre class="ql-syntax" spellcheck="false">discords_idx = np.full(l, -1, dtype=np.int64)
discords_dist = np.full(l, np.NINF, dtype=np.float64)
discords_nn_idx = np.full(l, -1, dtype=np.int64)

n = 0
while (np.any(include)):
    .
    .
    .
    if np.any(is_cands):
        # Assuming that we return multiple discords
&nbsp; &nbsp; &nbsp; &nbsp; IDX, D, I = _find_discord(T, m, M_T, Σ_T, is_cands, approx_P)
        for idx, d, i in zip(IDX, D, I):
            if n &lt; k:
&nbsp; &nbsp; &nbsp; &nbsp;         discords_idx[n] = idx
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;     discords_dist[n] = d
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;     discords_nn_idx[n] = i
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;     core.apply_exclusion_zone(include, idx, excl_zone, val=False)
            n += 1
</pre><p> </p><p>I guess, alternatively, we can over-append to the current lists and then remove any additional discords beyond <code>k</code> :</p><p> </p><pre class="ql-syntax" spellcheck="false">discords_idx = []
discords_dist = []
discords_nn_idx = []

n = 0
while (np.any(include)):
    .
    .
    .
    if np.any(is_cands):
        # Assuming that we return multiple discords
&nbsp; &nbsp; &nbsp; &nbsp; IDX, D, I = _find_discord(T, m, M_T, Σ_T, is_cands, approx_P)
&nbsp; &nbsp; &nbsp; &nbsp; discords_idx.extend(IDX)
&nbsp; &nbsp; &nbsp; &nbsp; discords_dist.extend(D)
&nbsp; &nbsp;     discords_nn_idx.extend(I)
        for i in IDX:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; core.apply_exclusion_zone(include, i, excl_zone, val=False)
        n += len(IDX)

discords_idx = discords_idx[:k]
discords_dists = discords_dists[:k]
</pre><p> </p><p>Hmmm, maybe this (using Python lists) is a little bit cleaner than going back to numpy arrays</p>

Reply:
> Regarding the edge case, yes, I agree with what you proposed. We should set `min_dist = 1e-6` if it ever drops below `config.STUMPY_MIN_DISCORD_DISTANCE`

I will change it in next set of commits. Also... I think `if min_dist < 1e-6: min_dist = 1e-6` is more readable compared to using `max`.  I think it better fits to what you said "... if it ever drops below...". 
**please let me know if you think otherwise**


> Yes. I'm starting to wonder if we should try to be consistent with the paper and simply call this `r` since `min_dist` makes it too ambiguous. And then borrow from the description in section B of the paper.

I totally agree. A user might also feel some sort of conflict between `max_dist` in `motifs` and `min_dist` in `discords` in STUMPY. So, I will change all `min_dist` to `r` and will use the description provided in the paper. I will revise it in next set of commits.

---

> Regarding n = 0 , I just meant something like:

I read your note in another comment regarding this matter. You want to keep updating index so, as you said, you can keep track of where you are in filling arrays. 

> Hmmm, maybe this (using Python lists) is a little bit cleaner than going back to numpy arrays

I have the same feeling. In that case, we do not need to use `indexer n`. Right? (as you mentioned: "just over-append and then remove any additional discords beyond k")

Reply:
<p>I agree with you. Let's just append</p>

Reply:
Btw, one alternative approach is to add new input `k_max` (with default 1) to `_find_discords` which tells function the maximum number of discords that should be returned from a set of candidates. So, while we are searching for discords in candidates in `_find_discords`, we can break when we get to `k_max` and return them. 

So, if we do:
```
while (np.any(include)):
   k_max = k - len(discords_idx)
   #some code
  if np.any(is_cands):
     #some code
     IDX, D, I = _find_discord(T, m, M_T, Σ_T, is_cands, approx_P, k_max = k_max)
   
``` 

Then, we will not over-append.

Reply:
<p>It's ok, it's not worth the added code complexity for now. Let's not do that</p>

Reply:
I will change the while-loop conditional expression as well as replace `min_dist` with `r` (throughout the notebook) in next set of commits.

Reply:
<p>I think <code>r</code> and <code>r_init</code> are fine  </p>

Reply:
Thanks!